onboard/catkin_ws/src/cv/scripts/depthai_sonar_spatial_detection.py

```diff
@@ -0,0 +1,463 @@
+#!/usr/bin/env python3
+
+import rospy
+import resource_retriever as rr
+import yaml
+import math
+import depthai_camera_connect
+import depthai as dai
+import numpy as np
+from utils import DetectionVisualizer
+from cv_bridge import CvBridge
+
+from custom_msgs.srv import EnableModel
+from custom_msgs.msg import CVObject
+from custom_msgs.msg import sweepResult, sweepGoal
+from sensor_msgs.msg import Image
+from geometry_msgs.msg import Pose
+
+
+MM_IN_METER = 1000
+DEPTHAI_OD_MODELS_FILEPATH = 'package://cv/models/depthai_models.yaml'
+HORIZONTAL_FOV = 95
+CAMERA_PIXEL_WIDTH = 416
+SONAR_DEPTH = 10
+SONAR_REQUESTS_PATH = 'sonar/request'
+SONAR_RESPONSES_PATH = 'sonar/cv/response'
+POSE_PUBLISHER_PATH = 'cv/pose'
+
+
+# Compute detections on live camera feed and publish spatial
+# coordinates for detected objects by aggregating sonar and stereo vision
+class DepthAISpatialDetector:
+    def __init__(self):
+        """
+        Initializes the ROS node and service. Loads the yaml file
+        at cv/models/depthai_models.yaml
+        """
+        rospy.init_node('depthai_sonar_spatial_detection', anonymous=True)
+
+        with open(rr.get_filename(DEPTHAI_OD_MODELS_FILEPATH,
+                                  use_protocol=False)) as f:
+            self.models = yaml.safe_load(f)
+
+        self.camera = 'front'
+        self.pipeline = None
+        self.publishers = None
+        self.output_queues = {}
+        self.current_model_name = None
+        self.classes = None
+        self.detection_feed_publisher = None
+        self.rgb_preview_publisher = None
+        self.detection_visualizer = None
+
+        self.enable_service = f'enable_model_{self.camera}'
+
+        self.bridge = CvBridge()
+        self.sonar_response = (0, 0)
+        self.in_sonar_range = True
+
+        self.sonar_requests_publisher = rospy.Publisher(
+            SONAR_REQUESTS_PATH, sweepGoal, queue_size=10)
+        self.sonar_response_subscriber = rospy.Subscriber(
+            SONAR_RESPONSES_PATH, sweepResult, self.update_sonar)
+
+        self.pose_publisher = rospy.Publisher(POSE_PUBLISHER_PATH,
+                                              Pose, queue_size=10)
+
+    def build_pipeline(self, nn_blob_path, sync_nn=True):
```

Comment:
Is this stuff copy and pasted from the non-sonar DepthAI code? If so, could we make this extend the other class, or at least extract the common code somewhere else? 
I was told to tag @ricohasgithub

Reply:
That's a good point. I think the code is slightly different as we took out some stuff with stereo that we won't need due to the use of sonar to help speed things up. We can definitely consider this in a future update.